### ISSUE TITLE
More efficiently expire data files

### ISSUE THREAD

**Issue by jiminoc** [dgraph-io/badger#562]
I'm looking at using badger for data that can all be TTL'd out uniformly. e.g. 1hr TTL on every key/val. 

Currently, I cannot get cleanup to keep up with write rate so badger will eventually fill up disk and kill the machine. I went through this same issue with Cassandra and the solution we got into place there was more efficient compaction where you store metadata on the SSTAble file that holds the earliest TTL and the latest TTL timestamp, if both of those are > than the TTL then you can just RM the whole file. The benefit there is you reduce the disk reads by not having to read 500MB+ files into memory, affecting pagecache and disk performance. 

Is that something that could be done in badger and where might the best place in the code be to look at adding something like that? 

So basically each vlog and sst file would have two metadata timestamps that the compactor can read to see if it can just delete the file vs scan it. 

My alternative is to shard badger DBs by time bucket which I'm less excited about.

**Comment by Kelindar** [comment]
We're currently facing the same issue, ingesting very short-lived data (1h TTL) into and it seems like badger just runs out of disk space and dies.

**Comment by manishrjain** [comment]
I'd recommend trying the `LSMOnlyOptions`, though I'd still set ValueLogLoadingMode to options.MemoryMap. See if that'd help make TTLs better, it would definitely allow faster reclamation of the value log files.

**Comment by Kelindar** [comment]
@manishrjain thanks, I think we tried with `ValueLogLoadingMode: options.MemoryMap`, also something that helped was to do GC every `1 minute` (but seems excessive to me). Not sure about `LSMOptionsOnly`, as our values are pretty big (16 MB). We're using a sharded DB which ingests around 1 TB per hour with 1h TTLs.

Would changing size of ValueLog files help at all? Anything else you can think of?

**Comment by campoy** [comment]
Conversation closed after lack of activity.
Please feel free to reopen if your concerns are still relevant.

**Comment by jiminoc** [comment]
@campoy is at DGraph now?!? sweet!!! Welcome aboard :)

**Comment by campoy** [comment]
Thanks, @jiminoc 😄
